### ISSUE TITLE
[Request] Web Interface for VNC and SPICE Configuration

### ISSUE THREAD

**Issue by BrainwreckedTech** [cockpit-project/cockpit-machines#553]
Page: Virtual Machines -> [your_virtual_machine]

Allow configuration of VNC and SPICE remote viewing through the web interface instead of requiring manual XML configuration.

If SPICE viewing is disabled, disable relevent SPICE devices in the XML config file.

Mockup:
![mockup_remote_display](https://user-images.githubusercontent.com/3752022/152339712-eb6a763e-eb8c-4a2b-b29e-53031fde9775.png)

**Comment by garrett** [comment]
Here's the desktop of CentOS Stream 9 in Cockpit Client (Cockpit's "Desktop" app):

![Screenshot from 2022-02-03 17-46-59](https://user-images.githubusercontent.com/10246/152389137-89ab93b0-928e-412d-bd4f-fb5e336456bf.png)

And the same VM connected via SPICE using "Virtual Machine Viewer" (virt-viewer):
![Screenshot from 2022-02-03 17-50-48](https://user-images.githubusercontent.com/10246/152389131-ab1b68a9-dfb2-4813-9bed-d553502d37cd.png)

But, yeah, thanks for opening this issue! We can make this more discoverable, and it should be editable as you suggest.

**Comment by BrainwreckedTech** [comment]
I didn't miss the info portion.  The Desktop Viewer part was actually really neat.

But it doesn't let you change anything.

In my case, I'm running libvirt VMs from a central home file server and would like to connect using a viewer on my desktop computer, and I'd like to _not_ have to edit an XML config file to do it.

![Screenshot_20220203_120155](https://user-images.githubusercontent.com/3752022/152391554-37111263-78ee-495a-8b71-d2a046a86618.png)

**Comment by garrett** [comment]
@BrainwreckedTech: Thanks for the quick reply!

I agree with you; you should be able to edit it without having to edit an XML.

I think it probably makes the most sense in the console area, as it's related and you wouldn't add multiples like the disks, network interface, etc.… The information is basically host, port, and user/password, so it's not too much information either. I think we can make it work.

(And if not there, perhaps in the overview?)

It's the end of the work week for me right now, but I have an idea and want to mock this up soon!

**Comment by BrainwreckedTech** [comment]
The console area would be fine if SPICE and VNC are the only options.  I just didn't want to assume that there weren't other options that the feature could be expanded to include.

I could also see people wanting to disable both options if a VM is considered live, only enabling a local (to the VM) VNC or SPICE option in emergency situations.

**Comment by Xenophilicy** [comment]
I second this idea, and I think the mockup that BrainwreckedTech threw together is gorgeous. I too would like to be able to add, remove, and update specific remote access protocols from the web interface instead of using `virsh edit` every time I need to adjust something. I'm glad this issue was opened so recently, as I just started to use cockpit a few days ago (I am hosting a few VMs for some friends) and was looking for this exact feature request.

**Comment by BrainwreckedTech** [comment]
I've been thinking things over, and between this request and my other one, I think my original mockup is the better way to go.  While the page might get long, it's better to have an overview of all the options you can set rather than tucking options away in miscellaneous spots to save space.

With a web interface, people are more accepting of having to scroll.  It's something that's expected when viewing things through a browser.  Even if someone doesn't understand the internals of HTML, CSS, and JavaScript, and how layout control is loosey-goosey on the web, they get the overall feeling of it through all of the web sites they visit throughout their lifetime.  So scrolling is no big deal.

Contrast this with desktop design, where precision is expected so we get endless debates on whether client-side decorations or the traditional title bar + menu bar + tool bar is the better use of space.  :)

**Comment by garrett** [comment]
Here's what I was thinking about having the remote options integrated into the console area:

![VM console remote options](https://user-images.githubusercontent.com/10246/154111038-33677933-1961-4166-8303-009c6157c4ab.png)

I don't have the edit modal dialog mocked up, but it would include both SPICE and VNC, probably both with a checkbox (so you can disable one or the other), with dropdowns to select the IP address (local only or one that can be accessed over a network) and a text entry (limited to numbers) as a port.

I'm currently unsure where the remote viewer button would go. Is it weird to have it in the popover? Otherwise, we could try to compress the line a bit more and have a button to launch remote or something like that.

Anyway, this mockup would certainly need more polishing; I wanted to make and share it ASAP to communicate my idea and so we have something more to talk about.

**Comment by BrainwreckedTech** [comment]
That also "unhides" the options.  :+1:

If Graphics/Text == VNC/Serial, I'd vote to stick with VNC/Serial.

What if "Remote" was a button, and the popup had modfied text from current layout?

> Clicking "Remote" will download a `.vv` file and launch Remote Viewer.  Search for "Remote Viewer" or `virt-viewer` in your software/package manager.  (For Windows, [download the MSI from virt-manager.org](https://virt-manager.org/download/))

**Comment by garrett** [comment]
@BrainwreckedTech: Thanks! Those are great suggestions.

I've adjusted the console on the Virtual Machine details page:

![VM console](https://user-images.githubusercontent.com/10246/157207848-27569e7d-5709-414d-9ac9-2c04945534b8.png)

And moved the remote viewer information in a dialog:

![Remote viewer confirm modal (2)](https://user-images.githubusercontent.com/10246/157209201-41c5d99e-756a-4c01-bb61-1c990da58bfa.png)

And I think it might get a bit tedious to set up the handling and have to do multiple click throughs, so I added skip setting to bypass the dialog for the next time. But then that needs a way to undo the setting. It's a browser-based setting that would be "global" to all VMs. Right now, we don't have any settings for machines, and if we did, it would be at the main page listing all the VMs, not in the details page.

So, while this is a browser setting that applies to all VMs, I think it still makes sense to expose the toggle close to where you can set it. As a compromise, I've added it to the edit SPICE and VNC settings dialog (which is new here — it was just implied previously):

![SPICE and VNC settings modal](https://user-images.githubusercontent.com/10246/157207852-a5d3b595-aa79-4a5a-948b-9b844297dcb6.png)

It's wrong to have a global feature (that's browser-based) at this level, but it's also wrong (and probably _even more wrong_) to have a global setting out of sight, so "perfect is the enemy of good" and all that... and right now, I think this is the best approach, even if it's a little bit of a compromise.

**Comment by Asleep-at-the-wheel** [comment]
> @BrainwreckedTech: Thanks! Those are great suggestions.

> ![SPICE and VNC settings modal](https://user-images.githubusercontent.com/10246/157207852-a5d3b595-aa79-4a5a-948b-9b844297dcb6.png)
> 
> It's wrong to have a global feature (that's browser-based) at this level, but it's also wrong (and probably _even more wrong_) to have a global setting out of sight, so "perfect is the enemy of good" and all that... and right now, I think this is the best approach, even if it's a little bit of a compromise.

Would it be possible to change those address input boxes to drop down boxes where a user could auto select any of the interfaces on the host?

For example, drop down lists all interfaces on the host dynamically like this:
 - virbr0
 - virbr1
 - lo
 - custom

Under custom, an input box shows up and the user can enter what ever IP they want, with the other options, the IP is pulled from the interfaces? Possibly mention somewhere that the firewall will need adjusting too.

**Comment by garrett** [comment]
Cockpit Client needs the connection to work over SSH. Perhaps we should have the connection happen through SSH anyway, otherwise the ports won't matter, unless the firewall is bypassed.

So we'd need to either:
- always use SSH
- instruct the firewall to temporarily open a port
- an option to switch (in some cases)

The simplest that should work virtually everywhere (that has SSH enabled at least, which is probably almost every machine that uses Cockpit) would be to just force the VNC and SPICE connections through SSH. With this being the case, the port selection doesn't make so much sense, right? And the connection would always be localhost once it's through SSH? This is all assuming that virt-viewer supports SSH in the vv file. (It does at the command line.)

Example virt-viewer connection over SSH:

```
virt-viewer --connect qemu+ssh://user@host:222/system guest
```

...but does big question is: the configuration file support this?

Meanwhile, our IP + port settings don't even work unless one (or more) is true:
- the firewall is down
- the firewall has a port poked through
- someone is running Cockpit on the same machine as using VNC

...which means the desktop connection is currently almost entirely useless by default. And this is why I'm suggesting we use ssh instead (or at least by default).

While the mockups I posted above would mainly work for this, I think we'd probably want to retool them a bit more based on what's decided here.

**Comment by Asleep-at-the-wheel** [comment]
You are correct and sticking to SSH to provide an encrypted tunnel is always preferred. Users can however change the XML of the VM to bind VNC to an IP/port on the host and open a firewall port. This allows users to connect with any VNC client from anywhere.

Probably want to somehow incorporate this use case into your draft (which I suggested using drop downs to bind to IP of interface, and perhaps defaults to **lo**).

**Comment by 1ras** [comment]
I'm currently evaluating Cockpit Machines and one of the things I stumbled upon is the Spice integration. Launching the viewer with just one click is really something I miss. 

In addition it would also be nice to launch the viewer with one click from the overview page (next to the Run/Shutdown buttons).

Regarding the "skip explanation dialog" proposal, that would not be a domain-related setting but a user-related setting (every user need to see the message once until he installed the Spice client). Perhaps a question mark with an explanation of the requirements when clicked would be an easier approach. Similar to the explanation of the different interface types of a network configuration.

I would have liked to test the SSH tunnel approach to compare its speed, but it doesn't seem to be supported by the Windows virt-viewer, which unfortunately is a must for the setup I'm evaluating.

In order to be able to use the viewer remotely via VPN, I had to enabled the Spice compression. So my suggestion would be an option for it in the settings.

I'm also a bit surprised why the spice_listen option in /etc/libvirt/qemu.conf is copied to each domain? If it's a default, there shouldn't be a need to copy it.

**Comment by garrett** [comment]
> You are correct and sticking to SSH to provide an encrypted tunnel is always preferred.

Right. We should default to SSH. The gotcha is that it seems the remote viewer app doesn't support SSH in the config file?

@jelly added a bug about this upstream @ https://gitlab.com/virt-viewer/virt-viewer/-/issues/86

> Users can however change the XML of the VM to bind VNC to an IP/port on the host and open a firewall port.

Editing an XML file and opening a firewall port for this really doesn't belong in Cockpit in such a manual way. (We're basically doing just part of that now, without any firewall manipulation or guidance. :frowning_face:) 

Cockpit _should_ make it easy to do, without any extra effort, and this includes opening a firewall port from the same place in Machines. We do have a firewall page for this, but it should be added from the Machines page, in this configuration area if `cockpit-firewalld` is installed. (Otherwise, it needs to provide information about opening a firewall port or possibly have an opt-in to use something like .)

Right now, we're "totally dropping the ball" on this feature, as the it really doesn't work anywhere without manual work, except on development machines. 

I think adding a desktop viewer is an old feature, probably even added before we had firewall support, which would help to explain this.

> Probably want to somehow incorporate this use case into your draft (which I suggested using drop downs to bind to IP of interface, and perhaps defaults to lo).

It's there already, as `instruct the firewall to temporarily open a port` and also implied in `an option to switch` (between SSH and the IP+port+firewall method), which is basically the IP + port of what we have now, with the addition of actually letting that port do something useful by Cockpit opening up the port in the firewall.

Since we already do have the IP and port UI and the new design relies on that for now, perhaps we should just finish it first, by opening up a port in the firewall. As this would automatically open up another means of access to the system (so it's actually useful from Cockpit), we'd need a warning that this will open up a firewall port.

And _then_, we would extend this by adding SSH and making it the default.

**Comment by garrett** [comment]
> Regarding the "skip explanation dialog" proposal, that would not be a domain-related setting but a user-related setting (every user need to see the message once until he installed the Spice client).

 Yes, it would likely be in the browser LocalStorage, which means it's a setting stored in the browser... which means per-user, effectively.

> In order to be able to use the viewer remotely via VPN, I had to enabled the Spice compression. So my suggestion would be an option for it in the settings.

We should probably have it enabled by default, unless there's a good reason not to? (I guess CPU overhead and latency might be of concern? But it's likely really negligible in practice on any modern machine.)

> I'm also a bit surprised why the spice_listen option in /etc/libvirt/qemu.conf is copied to each domain? If it's a default, there shouldn't be a need to copy it.

This feature has been around for a while in its current state. There might be a reason for it, or it could be due to an old default. Or possibly some distributions we support might not have it enabled by default. We should reevaluate this. Thanks!

**Comment by garrett** [comment]
Thanks again for all the feedback, everyone! Hopefully you'll still be happy to keep sharing your feedback as we work on this more.

I've added a ticket internally (to make this _officially_ higher priority), where I outline my current plan:

1. Improve web UI for VNC and serial connections
2. Improve UI for local desktop-based access (using a `.vv` file)
3. Add automatic firewall port opening for IP+port method (with a warning that this will open a port on the machine; a warning is needed as it increases the attack surface)
4. Add the ability to connect over SSH (and make it default)
5. Handle file transfers and launching in Cockpit Client

Some of these could be done out of order or in parallel. I'll need to adjust designs for the plan too (especially for SSH), but what I have above is probably still a good starting place.

(By-the-way, just to be clear: Despite this being _also_ tracked with our internal tracking tool, we'll still work in the open here as always, and hope to get more feedback along the way too as we share our progress. Our internal tracking tool is secondary to GitHub's issues, but mainly used for planning larger & important items. In other words, I'm basically using the internal tool to try to prioritize this issue — and I do link to and refer to our conversation here. :wink:)

**Comment by jelly** [comment]
From #virt on oftc I recieved an answer on what is the difference between `remote-viewer` and `virt-viewer`:

```
remote-viewer is a general purpose  vnc/spice client where you must tell it the address of the server to connect to. virt-viewer meanwhile is a specialization of remote-view, which takes a libvirt guest name, and connects to libvirt to auto-discover the VNC/spice server address.
```

**Comment by garrett** [comment]
The file handling part of Cockpit Client is being tracked in https://github.com/cockpit-project/cockpit/issues/17626. 

As Cockpit Client connects through SSH (not using/needing the Cockpit port of 9090), it's all the more important for users of Cockpit Client that we make sure remoting VNC/SPICE works via SSH instead of setting a port and opening it up in the firewall. (Logic being that if someone doesn't want or need port 9090 open, they probably also wouldn't want a SPICE or VNC port open either.)

In other words: SSH is important; we should do this somehow. If the client doesn't support it via a launcher file yet, then we might need to show the command line to make it work.

(However, the command line would only work in Linux, not in Windows and macOS. WSL2 would be an exception, but that's still Linux... just in Windows. Same issue, really, and we shouldn't expect people to have to install and set up WSL2 enough to get remote-viewer / virt-viewer working.)

**Comment by 1ras** [comment]
Regarding the Spice compression: Currently when enabling Spice compression with virsh edit or virt-xml, it's still disabled by virt-install during installation. It works as intended **after** installation but not during installation.

It seems that virt-install uses a logic which explicitly disables Spice compression when accessed locally and with Cockpit every access is local.

**Comment by AnterCreeper** [comment]
There should be a password setting, which is necessary for remote access.

**Comment by KKoukiou** [comment]
I really like the mockups here @garrett ! I will check how easy it's to implement. This will need to be done upstream in the @partternfly/react-console component.

**Comment by KKoukiou** [comment]
@garrett for the serial and vncconsole there used to be a button to 'Send keys'. Also there used to be a 'Disconnect' button. Dropping this will probably will be unsatisfying some users. Can we fit this into the design?

Also the instructions on how to install the virt-viewer for the remote viewer are missing in this mockup. Was this deliberate change?

**Comment by garrett** [comment]
> @garrett for the serial and vncconsole there used to be a button to 'Send keys'. Also there used to be a 'Disconnect' button. Dropping this will probably will be unsatisfying some users. Can we fit this into the design?

Sure, good point.

> Also the instructions on how to install the virt-viewer for the remote viewer are missing in this mockup. Was this deliberate change?

I thought they are in the mockup?

The first version was @ https://github.com/cockpit-project/cockpit-machines/issues/553#issuecomment-1040537199

The revised one moved it: https://github.com/cockpit-project/cockpit-machines/issues/553#issuecomment-1061581973:

> And moved the remote viewer information in a dialog:
>
> ![image](https://user-images.githubusercontent.com/10246/207325944-aa2396fc-134c-4b44-bf89-44d46d3fbb23.png)

**Comment by garrett** [comment]
>  @garrett for the serial and vncconsole there used to be a button to 'Send keys'. Also there used to be a 'Disconnect' button. Dropping this will probably will be unsatisfying some users. Can we fit this into the design?

This design is for the smaller version, BTW; the expanded version should have everything.

We could probably have everything in a kebab menu, like:

- Send key
   - all the keys underneath as options
- \--- divider \---
- Disconnection (as a dangerous link)

**Comment by renich** [comment]
Maybe setting the keyboard language as well?

**Comment by madwax** [comment]
Fooling issue should be linked to this #73

**Comment by martinpitt** [comment]
@garrett I have some questions wrt. the design [above](https://github.com/cockpit-project/cockpit-machines/issues/553#issuecomment-1061581973): 
 - So you want to pull out "Desktop" from the Select dropdown, and replace the rest (Serial/VNC) with two buttons *if* there is more than one console available, and otherwise with nothing, and just show the one available console?
 - The "Launch viewer" button would replace the "Desktop" select dropdown. So it would show the "Remote viewer" info dialog, and "Launch viewer" then triggers the .vv download? and checking "Skip next time" would set a localstorage flag and immediately download the .vv next time?
 - The edit button is missing a way to delete consoles. This is by far the highest priority aspect of this bug (see the linked RHEL issues). We can add a :wastebasket: icon next to them. But then we also need a way to add them back I suppose? That starts to sound complicated then, but we can't "disable" a console -- there's no storage/API in libvirt to do that.

**Comment by GreatBigWhiteWorld** [comment]
I had the issue (prolly #73) with spice server when setting up the new guest machine, so I chose 'replace spice server'.
After I found some 'workaround' for spice server I'd like to give it a try, but I can't find any option to use spice server again. Is it related to this issue (of not possible to add console)?
Any tip to get back to use spice server for now?

**Comment by mvollmer** [comment]
I am restarting work on this. I'll track it here: https://issues.redhat.com/browse/COCKPIT-868

I would like to propose the following:

 - We ignore Spice completely in Cockpit. Cockpit has no inline viewer for it, we will likely not let people add it or edit it, it's deprecated on RHEL, and the "Launch viewer" button doesn't really work either. Cockpit not supporting Spice doesn't hurt much. This makes it easier for Cockpit to drop the "Launch viewer" button and keep the UI streamlined. (The mockups mention Spice in the VNC console tab, for example. Awkward.)

 -  We have this in the VNC footer: "VNC listening on <ip>:<port> — Edit --- How to connect" The "How to connect" button pops up an explanation of how to install virt-viewer and how to run it over SSH.
 
 -  There is no "Launch viewer" button anymore.

I will take the dialog implementations from #1795.

**Comment by mvollmer** [comment]
> Any tip to get back to use spice server for now?

`virt-xml --add-device --graphics spice <vmname>` worked for me.

**Comment by martinpitt** [comment]
> We ignore Spice completely in Cockpit

Good reasoning if you mean "Spice console", fine for me (there's other spice funtionality like hardware passthrough or so, not sure how much we depend on that). However, please retain the "Remove/Convert SPICE devices" functionality.

**Comment by mvollmer** [comment]
> Good reasoning if you mean "Spice console", fine for me

Yes, I am only talking about the "Consoles" card context here.

> However, please retain the "Remove/Convert SPICE devices" functionality.

Yes, of course.

**Comment by 1ras** [comment]
> I am restarting work on this. I'll track it here: https://issues.redhat.com/browse/COCKPIT-868
> 
> I would like to propose the following:
> 
>     * We ignore Spice completely in Cockpit. Cockpit has no inline viewer for it, we will likely not let people add it or edit it, it's deprecated on RHEL, and the "Launch viewer" button doesn't really work either. Cockpit not supporting Spice doesn't hurt much. This makes it easier for Cockpit to drop the "Launch viewer" button and keep the UI streamlined. (The mockups mention Spice in the VNC console tab, for example. Awkward.)
> 
>     * We have this in the VNC footer: "VNC listening on `<ip>:<port>` — Edit --- How to connect" The "How to connect" button pops up an explanation of how to install virt-viewer and how to run it over SSH.
> 
>     * There is no "Launch viewer" button anymore.
> 
> 
> I will take the dialog implementations from [#1795](https://github.com/cockpit-project/cockpit-machines/pull/1795).

Here is some feedback of my use case:

I use the "Launch remote viewer" button daily because VNC is nice as a preview and to take a quick look, but misses the "native" view of the guest. E.g. Mouse clicks end in the wrong place if the browser window is not big enough and the view is scaled (bug #981), too much space consuming as VNC is always surrounded by the cockpit layout (bug #965), no resize with automatic change of the guest resolution, no full-screen with resolution change of the guest, no virtual USB.

Setting up Spice is not well supported within Cockpit, but once manually setup, the "Launch remote viewer" button is still very useful.

The "Launch remote viewer" button is currently two mouse clicks away where the workaround with the SSH tunnel is so far away, that I would not even consider it as a feasible workaround.

**Comment by mvollmer** [comment]
> The "Launch remote viewer" button is currently two mouse clicks away where the workaround with the SSH tunnel is so far away, that I would not even consider it as a feasible workaround.

Thanks for the feedback! Let's try to keep your use case working.

Can you tell us more about how you use this exactly?  Where does the Browser run, to which address are you connecting with the browser, is there SSH involved, are you using Spice or VNC, did you have to configure the Spice or VNC listening address/port, did you open the Firewall, etc?

For example, I know that the "Launch viewer" button works out of the box when the VM host is the same machine as the one running the Browser.

It will also work when the machine running the browser and the VM host are different, but the listening port is set to a IP address that is reachable from the browser machine and the firewall has been opened.

**Comment by 1ras** [comment]
I use it over network for a lab/test environment which has his own VPN. Therefore encryption would be nice to have but is not a must. I cannot use SSH tunnel as this would cut of Windows clients (my understanding is that virt-viewer on Windows doesn't support it). I tried Spice built-in encryption, but Cockpit is not prepared for it (not covered in the .vv file) and I didn't want to spend time with the Cockpit build process which sounds not that simple.

The spice listen address is setup in /etc/libvirt/qemu.conf and every new VM gets a default configuration automatically applied with a shell scrit that waits with inotify for new VMs. For spice it basically enables the spice image compression. The browser and also virt-viewer when started by the browser connects to the IP address of the KVM host and the firewall is open.

**Comment by mvollmer** [comment]
> I use it over network for a lab/test environment which has his own VPN. Therefore encryption would be nice to have but is not a must. I cannot use SSH tunnel as this would cut of Windows clients (my understanding is that virt-viewer on Windows doesn't support it).

I see, hmm. I was hoping we could make "virt-viewer -c qemu+ssh://" the gold standard of connecting, but maybe we should concentrate on using TLS for the VNC encryption. This should also work out of the box with the .vv files: https://issues.redhat.com/browse/COCKPIT-1242

> I tried Spice built-in encryption, but Cockpit is not prepared for it (not covered in the .vv file) and I didn't want to spend time with the Cockpit build process which sounds not that simple.

I have made a ticket to improve that in Cockpit: https://issues.redhat.com/browse/COCKPIT-1241
 
> The spice listen address is setup in /etc/libvirt/qemu.conf 

We have just merged #1982 which changes the address Cockpit uses in the .vv file. It used to be the listening address, but now it is the same address that Cockpit uses to reach the KVM host. This should be released with [Cockpit 332](https://jelly.github.io/cockpit-project.github.io/blog/cockpit-332.html). Now you should be able to set the listening address to 0.0.0.0. Does that make things easier for you?

> The browser and also virt-viewer when started by the browser connects to the IP address of the KVM host and the firewall is open.

Thanks! We are reluctant to let people open up the listening port in Cockpit itself unless we also support some form of encryption. But I now understand that we should keep the "Launch viewer" button for people that do their own "low level" plumbing to make it all work. So thanks for the feedback!

Why do you use Spice? I know Cockpit puts Spice into the .vv file if there is both VNC and Spice configured for a machine. But Cockpit itself only has a inline viewer for VNC and Spice is being deprecated in RHEL. We might not want to put a lot of effort into configuring Spice inside Cockpit, for those reasons. 

Do you also have VNC configured? Would it work for you if Cockpit would prefer VNC over Spice in the .vv file if both are configured?

**Comment by martinpitt** [comment]
Current cockpit-machines does that now, done in #2008.
